pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+
+```jsx
+const Hexagon = props => {
+  const renderingContext = React.useContext(SharingContext);
+  const frameCount = React.useContext(FrameContext);
+  // drawing logic
+};
+```
+
+And now our hexagons are back on the screen! 🎉 While this works, the `FrameContext` has to be added in every child component. Two options are available to make sure those two contexts are being consumed together:
+
+- We can regroup them in a single context, sharing both the `renderingContext` and the `frameCount`. But the `frameCount` variable is not used in the child components, so it does not make sense to share its value.
+
+- or, we can create a `useCanvas` hook to hide this complexity away! Even when consuming both React contexts, the hook can only return the canvas rendering context to the child components:
+
+```jsx
+export const useCanvas = () => {
+  React.useContext(FrameContext);
+  const renderingContext = React.useContext(CanvasContext);
+  return renderingContext;
+};
+```
+
+The `Hexagon` component logic now needs a small update to use this new hook:
+
+```jsx
+const Hexagon = props => {
+  const renderingContext = useCanvas();
+  // drawing logic
+};
+```
+
+### 🔄 Making the hexagons move
+
+We want the hexagons to rotate. It means that each hexagon should have its rotation angle change at every frame, by incrementing it by 1 for example. To do this, we need the hexagons to _remember_ the rotation from the previous render: we will need to use a [React ref](/react-refs-evolution):
+
+```jsx
+const animatedRotation = React.useRef(props.rotation);
+animatedRotation.current = animatedRotation.current + 1;
+```
+
+As the `Hexagon` component re-renders at every frame, this code makes its rotation  change at every frame: they rotate! 😍
+
+Like we did with `useCanvas`, we can also improve a bit the readability of this code. We can hide the implementation detail – here, the use of a React ref – in a hook:
+
+```jsx animated-hexagons-v9j2q src/Canvas.js
+const useAnimation = (initialValue, valueUpdater) => {
+  const animatedValue = React.useRef(initialValue);
+  animatedValue.current = valueUpdater(animatedValue.current);
+  return animatedValue.current;
+};
+```
+
+The `Hexagon` code now looks a bit better:
+
+```jsx animated-hexagons-v9j2q src/Hexagon.js
+const Hexagon = props => {
+  // [...]
+  const animatedRotation = useAnimation(props.rotation, angle => angle + 1);
+  // drawing logic
+};
+```
+
+👏 Congratulations 👏 We now have animated canvas-based React components! We even created two custom hooks along the way to make our code nicer.
+
+## 👀 Going further
+
+- At some point it can become quite CPU heavy to try to animate a lot of shapes in a canvas. At this point, there's another more performant way to draw things: WebGL. It's a huge subject on its own! I recommend to look at how we can write WebGL based components with [react-three-fiber](https://github.com/react-spring/react-three-fiber) _— spoiler alert, they made their own React reconciler_
```

Comment:
```suggestion
- At some point it can become quite CPU heavy to try to animate a lot of shapes in a canvas. At this point, there's another more performant way to draw things: WebGL. It's a huge subject on its own! I would recommend you to look at how to write WebGL based components with [react-three-fiber](https://github.com/react-spring/react-three-fiber) _— spoiler alert, they made their own React reconciler_
```